### PR TITLE
pageserver: exclude gc_horizon from synthetic size calculation

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

See:
- https://github.com/neondatabase/neon/issues/6374

## Summary of changes

Whereas previously we calculated synthetic size from the gc_horizon or the pitr_interval (whichever is the lower LSN), now we ignore gc_horizon and exclusively start from the `pitr_interval`.  This is a more generous calculation for billing, where we do not charge users for data retained due to gc_horizon.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by arpad-m — APPROVED** [review]

File: pageserver/src/tenant/size.rs [hunk 1]
```diff
@@ -183,7 +183,13 @@ pub(super) async fn gather_inputs(
...
         // we don't consider the `Timeline::disk_consistent_lsn` at all, because we are not
         // actually removing files.
-        let mut next_gc_cutoff = cmp::min(gc_info.horizon_cutoff, gc_info.pitr_cutoff);
+        //
+        // We only consider [`GcInfo::pitr_cutoff`], and not [`GcInfo::horizon_cutoff`], because from
+        // a user's perspective they have only requested retention up to the time bound (pitr_cutoff), rather
+        // than a space bound (horizon cutoff).  This means that if someone drops a database and waits for their
+        // PITR interval, they will see synthetic size decrease, even if we are still storing data inside
+        // horizon_cutoff.
+        let mut next_gc_cutoff = gc_info.pitr_cutoff;
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:192]: Weird that none of the tests failed. Creating a test which reproduces the problem seems quite straightforward. I'd run the disabled ones as well if any of them already are sensitive.

Comment (koivunej) [reply pageserver/src/tenant/size.rs:192]: Luckily even the ignored tests in `test_runner/regress/test_tenant_size.py` work on pg14 before this change.

Comment (koivunej) [reply pageserver/src/tenant/size.rs:192]: Now that we have a solution for reproducable initdb_lsn one could reconsider adding tighter values there as well, but of course there is the caveat of ignored but still increasing `last_record_lsn` cases which are probably a big chunk of the instability already.

Comment (jcsp) [reply pageserver/src/tenant/size.rs:192]: I finally got back around to this.  Good flag that there were tests that really ought to have failed.

I hadn't realized the synthetic size tests were in such a degraded state: opened https://github.com/neondatabase/neon/issues/7138 to track that.

`test_single_branch_get_tenant_size_grows` has a case for dropping a table, which checks that size _increases_, but it didn't cover this change because the test was running with a zero PITR interval.

Reinstating and extending that test to use a 1ms PITR interval to exercise this PR, I found that `update_gc_info` has a bug (I think).  If the PITR time point is more recent than the most recent commit, then it falls back to using the size-based GC horizon -- that doesn't make sense: if the PITR time point is more recent than the recent commit, it should be interpreted that the PITR LSN is the head of the log.  I've added a commit that changed that behavior.

Comment (koivunej) [reply pageserver/src/tenant/size.rs:192]: Yes, we added those extra tests back then but didn't leave them running because other tests should cover the cases. Those at least made it more straightforward back then to consider different situations.

> If the PITR time point is more recent than the most recent commit, then it falls back to using the size-based GC horizon -- that doesn't make sense: if the PITR time point is more recent than the recent commit, it should be interpreted that the PITR LSN is the head of the log. I've added a commit that changed that behavior.

I cannot imagine where the two would not be combined by `min`, before this PR that is.

File: test_runner/regress/test_tenant_size.py [hunk 2]
```diff
@@ -492,24 +505,41 @@ def get_current_consistent_size(
...
-
         prev_size = collected_responses[-1][2]
-
         check_size_change(current_lsn, initdb_lsn, gc_horizon, size, prev_size)
 
+        # Set a tiny PITR interval to allow the DROP to impact the synthetic size
+        # Because synthetic size calculation uses pitr interval when available,
+        # when our tenant is configured with a tiny pitr interval, dropping a table should
+        # cause synthetic size to go down immediately
+        tenant_config["pitr_interval"] = "1ms"
```
Comment (koivunej) [inline test_runner/regress/test_tenant_size.py:528]: Aah yes, but this has to be non-zero for this to work.

Comment (koivunej) [reply test_runner/regress/test_tenant_size.py:528]: If update_gc_info would also return `last_record_lsn` in the case where `pitr_period == Duration::ZERO`, then this would work as expected using pitr=0.

**Review by koivunej — APPROVED** [review]

File: test_runner/regress/test_tenant_size.py (hunk 2 above)
Comment (koivunej) [reply test_runner/regress/test_tenant_size.py:528]: Discussed, this would had changed the value for existing configurations.
